DrawerViewExample/DrawerViewExample/DrawerView.swift

```diff
@@ -0,0 +1,285 @@
+}
+
+@available(iOS 10, *)
+// swiftlint:disable:next type_body_length we will revisit the internals of this class when adding more functionality
+open class DrawerView: UIView {
+  public let peekView: UIView
+  private let contentView: UIView
+  private let scrollingContentView = UIScrollView(frame: .zero)
+
+  private lazy var panToExpandGestureRecogniser = UIPanGestureRecognizer(target: self, action: #selector(collapsedFilterDidPan))
+  private lazy var panToCollapseGestureRecogniser = UIPanGestureRecognizer(target: self, action: #selector(expandedFilterDidPan))
+  private var initialTouchPosition: CGFloat = 0
+  // swiftlint:disable:next implicitly_unwrapped_optional we always create it before using
+  private var propertyAnimator: UIViewPropertyAnimator!
+  // swiftlint:disable:next implicitly_unwrapped_optional the constraint is mandatory for this class
+  public var animatingConstraint: NSLayoutConstraint!
+  public weak var delegate: DrawerViewDelegate?
+
+  public init(peekView: UIView, contentView: UIView) {
+    self.peekView = peekView
+    self.contentView = contentView
+    super.init(frame: .zero)
+
+    addSubview(peekView)
+    setupPeekView()
+
+    addSubview(scrollingContentView)
+
+    scrollingContentView.backgroundColor = .white
+    scrollingContentView.topToBottom(of: peekView)
+    scrollingContentView.leftToSuperview()
+    scrollingContentView.rightToSuperview()
+    scrollingContentView.bottomToSuperview()
+
+    scrollingContentView.addSubview(contentView)
+    contentView.edgesToSuperview()
+    contentView.width(to: self)
+  }
+
+  private func setupPeekView() {
+    peekView.edgesToSuperview(excluding: .bottom)
+
+    let tapGestureRecogniser = UITapGestureRecognizer(target: self, action: #selector(filterDidTap))
+    peekView.addGestureRecognizer(tapGestureRecogniser)
+    peekView.addGestureRecognizer(panToExpandGestureRecogniser)
+    peekView.addGestureRecognizer(panToCollapseGestureRecogniser)
+    panToCollapseGestureRecogniser.isEnabled = false
+    peekView.isUserInteractionEnabled = true
+  }
+
+  public required init?(coder aDecoder: NSCoder) {
+    fatalError("init(coder:) has not been implemented")
+  }
+
+  @objc
+  func filterDidTap(_ gestureRecogniser: UITapGestureRecognizer) {
+    if isExpanded {
+      collapse()
+    } else {
+      expand()
+    }
+  }
+
+  public func expand() {
+    delegate?.filterViewWillExpand(self)
+    animatorForExpandGesture().startAnimation()
+  }
+
+  public func collapse() {
+    // this prevents us crashing under a circumstance which I can't remember
+    if animatingConstraint != nil {
+      animatorForCollapseGesture().startAnimation()
+    }
+  }
+
+  var upperAnimationLimit: CGFloat {
+    // This is a bit of a guess, could do with re-working. This prevents the filter going behind
+    // the navigation bar, which means you can never close the filters.
+    return -1 * min(scrollingContentView.contentSize.height, UIScreen.main.bounds.height - 300)
```

Comment:
this could be improved by having a "greater than" constraint, which it can't go above.